### ISSUE TITLE
Prefer a 'Add' call on HashSet<T> over 'Add' call guarded by 'Contains'

### ISSUE THREAD

**Issue by drieseng** [dotnet/roslyn-analyzers#6506]
I suggest to add a rule that encourages the use of the return value of `HashSet<T>.Add(T item)` for control flow over a call to `HashSet<T>.Add(T item)` guarded by a call to `HashSet<T>.Contains(T item)`.

I've seen code in the wild that uses `HashSet<T>.Contains(T item)` to check if an item already exists in the hashset, and - if it doesn't exists - uses `HashSet<T>.Add(T item)` to add the item:

```cs
var hashSet = new HashSet<string>();

if (!hashSet.Contains("A"))
{
    _ = hashSet.Add("A");
    return 1;
}

return 0;
```

I propose to have a rule that detects such code, and encourages users to use `HashSet<T>.Add(T item)` directly and use the return value of this method for control flow:

```cs
var hashSet = new HashSet<string>();

return hashSet.Add("A") ? 1 : 0;
```

**Comment by mpidash** [comment]
There is a rule that covers guarded set calls: [CA1868](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1868).

Although in your example it does not warn, as it currently does not handle cases where the return value is explicitly discarded.
It works for expression statements though, see:

https://github.com/dotnet/roslyn-analyzers/blob/22a5b5af1a402fbba34dfbbdeadeb5aa571d008e/src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Performance/DoNotGuardSetAddOrRemoveByContainsTests.cs#L509-L530

I think we should also handle the case in your example and treat it the same as the one above.
